### PR TITLE
IGNITE-20574 Sql. Do not permit to create tables in reserved system schemas

### PR DESCRIPTION
**Pull request by zstan**
Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManagerImpl.java [hunk 1]
```diff
@@ -603,4 +603,9 @@ private static class ParentIdAwareDescriptor<T> {
             this.id = id;
         }
     }
+
+    /** List of reserved schema names. */
+    public static List<String> reservedSchemas() {
+        return List.of(SYSTEM_SCHEMA_NAME);
+    }
```
Comment (AMashenkov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManagerImpl.java:610]:
```suggestion
    public static boolean isSystemSchema(String schemaName) {
        return List.of(SYSTEM_SCHEMA_NAME).contains(schemaName));
    }
```
Also, a list can be cached.
I'd moved the method to CatalogUtils or made it non-static.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/commands/AbstractTableCommand.java [hunk 2]
```diff
@@ -34,13 +36,17 @@ public abstract class AbstractTableCommand implements CatalogCommand {
     protected final String tableName;
 
     AbstractTableCommand(String schemaName, String tableName) throws CatalogValidationException {
-        this.schemaName = schemaName;
+        this.schemaName = schemaName != null ? schemaName.toUpperCase() : null;
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/commands/AbstractTableCommand.java]: done, thanks

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManagerImpl.java (hunk 1 above)
Comment (zstan) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManagerImpl.java]: done

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/commands/CatalogUtils.java [hunk 3]
```diff
@@ -138,6 +139,13 @@ public class CatalogUtils {
         ALTER_COLUMN_TYPE_TRANSITIONS.put(ColumnType.DECIMAL, EnumSet.of(ColumnType.DECIMAL));
     }
 
+    public static final List<String> SYSTEM_SCHEMAS = List.of(SYSTEM_SCHEMA_NAME);
+
+    /** System schema names. */
+    static boolean isSystemSchema(String schemaName) {
+        return SYSTEM_SCHEMAS.contains(schemaName);
```
Comment (AMashenkov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/commands/CatalogUtils.java:146]: Do we want to make schema name case insensitive by default?
If so, then it make sense to transform 'schemaName' to upperCase or lowerCase at first.

Comment (zstan) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/commands/CatalogUtils.java:146]: but it schemas uppercased, check https://github.com/apache/ignite-3/pull/3081/files#diff-afb2cc57b87013df26077689351c7b537c1638946ba0828358b7d3296e9e2c0fR327

Comment (AMashenkov) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/commands/CatalogUtils.java:146]: ok.
Do we allow create schema with name quoted name "'Schema'"?

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/commands/AbstractTableCommand.java [hunk 4]
```diff
@@ -41,6 +42,10 @@ public abstract class AbstractTableCommand implements CatalogCommand {
     }
 
     private void validate() {
+        if (schemaName != null && CatalogUtils.isSystemSchema(schemaName)) {
+            throw new CatalogValidationException(format("Operations with reserved schemas "
+                    + "keywords are not allowed, schema: {}", schemaName));
```
Comment (lowka) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/commands/AbstractTableCommand.java:47]: I think there is a typo here.

>Operations with reserved schemas keywords are not allowed, schema: 

The word `keywords` is unnecessary.

Comment (zstan) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/commands/AbstractTableCommand.java]: done

**Review by lowka — APPROVED** [review]

**Review by AMashenkov — APPROVED** [review]

**Pull request merged**